binary_search_tree/tree.py

```diff
@@ -14,35 +17,110 @@ class Tree:
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        current = self.root
+        
+        while current != None:
+            if current.key == key:
+                return current.value
+            elif key < current.key:
+                current = current.left
+            else:
+                current = current.right
+            
+
+        return None
+
 
-    # Time Complexity: 
-    # Space Complexity: 
+    def inorder_helper(self, current_node, items):
+        if current_node is not None:
+            self.inorder_helper(current_node.left, items)
+            items.append({"key": current_node.key, "value": current_node.value})
+            self.inorder_helper(current_node.right, items)
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        # visit all nodes in left, then root, then all nodes in right
+        items = []
+        self.inorder_helper(self.root, items)
+        # current = self.root
+        # if current:
+        #     items = self.inorder(current.left)
+        #     items.append(current.value)
+        #     items = items + self.inorder(current.right)
+        return items
+
+    def preorder_helper(self, current_node, items):
+        if current_node is not None:
+            items.append({"key": current_node.key, "value": current_node.value})
+            self.preorder_helper(current_node.left, items)
+            self.preorder_helper(current_node.right, items)
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)     
     def preorder(self):
-        pass
+        # visit root node, then left nodes, then right nodes 
+        items = []
+        self.preorder_helper(self.root, items)
+
+        return items 
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def postorder_helper(self, current_node, items):
+        if current_node is not None:
+            self.postorder_helper(current_node.left, items)
+            self.postorder_helper(current_node.right, items)
+            items.append({"key": current_node.key, "value": current_node.value})
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)    
     def postorder(self):
-        pass
+        # visit all nodes on left, then all nodes on right, then root node 
+        items = []
+        self.postorder_helper(self.root, items)
+
+        return items 
+    
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def height_helper(self, current_node):
+        if current_node is not None:
+            height_left = self.height_helper(current_node.left)
+            height_right = self.height_helper(current_node.right)
+            return (max(height_left, height_right) +1)
+        else:
+            return 0 
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
```

Comment:
This is true if it's not balanced! Otherwise, it is O(log n) if it _is_ balanced